File: character-sheet-companion-server/src/listeners/rolls/rollInitiativeListener.ts
import { Server, Socket } from "socket.io";
import {
  getFoundrySocketFromRoom,
  getIosSocketFromRoom,
} from "../../utilities/SocketUtilities";
import { SOCKET_EVENTS } from "../../constants/events";
import { InitiativeRoll } from "../../interfaces/rolls/InitiativeRoll";

export const rollInitiativeListener = (io: Server, socket: Socket): void => {
  const requestFoundryInitiativeRoll = (initiativeRoll: InitiativeRoll) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ROLL.REQUEST.REQUEST_FOUNDRY_INITIATIVE_ROLL,
        initiativeRoll,
        socket.id
      );
    }
  };

  const receiveFoundryInitiativeRoll = (
    initiativeRoll: InitiativeRoll,
    iosSocketId: string
  ) => {
    const iosSocket = getIosSocketFromRoom(io, socket, iosSocketId);
    if (iosSocket) {
      iosSocket.emit(
        SOCKET_EVENTS.SERVER.ROLL.SEND.SEND_FOUNDRY_INITIATIVE_ROLL,
        JSON.stringify(initiativeRoll)
      );
    }
  };

  socket.on(
    SOCKET_EVENTS.IOS.ROLL.REQUEST_FOUNDRY_INITIATIVE_ROLL,
    requestFoundryInitiativeRoll
  );
  socket.on(
    SOCKET_EVENTS.FOUNDRY.ROLL.SEND_INITIATIVE_ROLL,
    receiveFoundryInitiativeRoll
  );
};

File: CharacterSheetCompanion/CharacterSheetCompanion/Services/SocketListeners/RollListeners/RollInitiativeListener.swift
//
//  InitiativeListener.swift
//  CharacterSheetCompanion
//
//  Created by Zachary Sanders on 8/2/21.
//

import Foundation
import SocketIO

class RollInitiativeListener: SocketListener {
    let socket: SocketIOClient
    let jsonEncoder: JSONEncoder

    var initiativeRollCallback: ((InitiativeRollModel) -> Void)?

    init(socket: SocketIOClient) {
        self.socket = socket
        jsonEncoder = JSONEncoder()
    }

    func addSocketHandlers() {
        socket.on(SocketEvents.SERVER.ROLL.SEND.SEND_INITIATIVE_ROLL) { data, _ in
            do {
                try self.initiativeRollCallback?(SocketListenerUtility.parseSocketEventData(data))
            } catch let FoundryJSONError.errorMessage(errorMessage) {
                print(errorMessage)
            } catch {
                print(error)
            }
        }
    }

    func rollInitiative(initiativeRoll: InitiativeRollModel, completionHandler: @escaping (InitiativeRollModel) -> Void) {
        do {
            let jsonData = try jsonEncoder.encode(initiativeRoll)
            if let json = String(data: jsonData, encoding: .utf8) {
                socket.emit(SocketEvents.IOS.ROLL.REQUEST_FOUNDRY_INITIATIVE_ROLL, json)
                initiativeRollCallback = completionHandler
            }
        } catch {}
    }
}

extension SocketEvents.IOS.ROLL {
    static let REQUEST_FOUNDRY_INITIATIVE_ROLL = "ios:requestFoundryInitiativeRoll"
}

extension SocketEvents.SERVER.ROLL.SEND {
    static let SEND_INITIATIVE_ROLL = "foundry:sendInitiativeRoll"
}
